blockprod/src/detail/mod.rs

```diff
@@ -47,158 +85,291 @@ impl BlockProduction {
+                );
+                return Err(BlockProductionError::TipChanged(
+                    tip_at_start.block_id(),
+                    tip_at_start.block_height(),
+                    current_tip_index.block_id(),
+                    current_tip_index.block_height(),
+                ));
+            }
 
-        block_production.start().expect("Error starting Block Builder");
+            // TODO: instead of the following static value, look at
+            // self.chain_config for the current block reward, then send
+            // it to self.reward_destination
+            let block_reward = BlockReward::new(vec![]);
 
-        let recv = timeout(Duration::from_millis(1000), builder_rx.recv());
+            // TODO: see if we can simplify this
+            let transactions = match transactions_source.clone() {
+                TransactionsSource::Mempool => {
+                    self.collect_transactions().await?.transactions().clone()
+                }
+                TransactionsSource::Provided(txs) => txs,
+            };
+
+            let block_body = BlockBody::new(block_reward, transactions);
+
+            let tx_merkle_root = calculate_tx_merkle_root(&block_body)
+                .map_err(BlockCreationError::MerkleTreeError)?;
+            let witness_merkle_root = calculate_witness_merkle_root(&block_body)
+                .map_err(BlockCreationError::MerkleTreeError)?;
+
+            let block_header = BlockHeader::new(
+                current_tip_index.block_id(),
+                tx_merkle_root,
+                witness_merkle_root,
+                timestamp,
+                consensus_data,
+            );
+
+            let (result_sender, mut result_receiver) = oneshot::channel();
+            {
+                let chain_config = Arc::clone(&self.chain_config);
+                let current_tip_height = current_tip_index.block_height();
+                let stop_flag = Arc::clone(&stop_flag);
+                self.mining_thread_pool.spawn(move || {
+                    let block_header = Self::solve_block(
+                        chain_config,
+                        block_header,
+                        current_tip_height,
+                        stop_flag,
+                    );
+                    result_sender
+                        .send(block_header)
+                        .expect("Failed to send block header back to main thread");
+                });
+            }
 
-        tokio::select! {
-            msg = recv => match msg.expect("Block Builder timed out").expect("Error reading from Block Builder") {
-                BlockBuilderControlCommand::Start => {},
-                _ => panic!("Invalid message received from Block Builder"),
+            tokio::select! {
+                _ = &mut cancel_receiver => {
+                    // TODO: test cancellations
+                    stop_flag.store(true, std::sync::atomic::Ordering::SeqCst);
```

Comment:
Maybe my understanding is incorrect, but I thought that when you use `SeqCst` ordering in one place, a happens-before relationship is only created when *all* accesses to that same variable are `SeqCst` too... however there is a `Relaxed` access elsewhere in the code.

Reply:
The right thing to do here is acquire-release, but since it's a one-time thing, it doesn't matter.

Usually we got for relaxed when performance is an issue.